io_scene_godot/converters/material_node_tree.py

```diff
@@ -0,0 +1,1090 @@
+
+    in_arguments = list()
+    in_arguments[0] = in_vec
+    in_arguments[1] = mat
+    in_arguments[2] = clamp_min
+    in_arguments[3] = clamp_max
+    in_arguments[4] = use_min
+    in_arguments[5] = use_max
+
+    out_vec = shader_script.create_variable(
+        "vec3", node.name, "mapping_vector"
+    )
+    shader_script.define_variable(out_vec)
+
+    add_function_call(
+        global_script, shader_script, function, in_arguments, [out_vec]
+    )
+
+    shader_script.socket_to_var_map[node.outputs[0]] = out_vec
+
+
+def parse_converter_node(global_script, shader_script, node):
+    """Traversal the node tree down to the leaf and convert each node
+    to a function call"""
+    function = find_node_function(node)
+    in_arguments = list()
+
+    for socket_id in function.in_sockets:
+        socket = node.inputs[socket_id]
+        if socket.is_linked:
+            # iput socket only has one link
+            link = socket.links[0]
+            if link.from_socket not in shader_script.socket_to_var_map:
+                parse_non_shader_node(
+                    global_script, shader_script, link.from_node
+                )
+            in_arguments.append(
+                shader_script.socket_to_var_map[link.from_socket]
+            )
+        else:
+            in_arguments.append(
+                Value.create_from_blender_value(socket.default_value)
+            )
+
+    out_arguments = list()
+    for index, socket in enumerate(node.outputs):
+        new_var = shader_script.create_variable(
+            function.out_param_types[index], node.name, socket.identifier
+        )
+        shader_script.socket_to_var_map[socket] = new_var
+        shader_script.define_variable(new_var)
+        out_arguments.append(new_var)
+
+    add_function_call(
+        global_script, shader_script, function, in_arguments, out_arguments
+    )
+
+
+def bump_displacement(displacement_link, global_script, fragment_script):
+    """Displacement as bump in fragment shader"""
+    parse_non_shader_node(
+        global_script,
+        fragment_script,
+        displacement_link.from_node
+    )
+    disp_output = fragment_script.socket_to_var_map[
+        displacement_link.from_socket
+    ]
+    fragment_script.add_bump_displacement(global_script, disp_output)
+
+
+def true_displacement(displacement_link, global_script, vertex_script):
+    """Real displacement in vertex shader"""
+    parse_non_shader_node(
+        global_script,
+        vertex_script,
+        displacement_link.from_node,
+    )
+    disp_output = vertex_script.socket_to_var_map[
+        displacement_link.from_socket
+    ]
+    vertex_script.add_displacement(global_script, disp_output)
+
+
+def export_node_tree(escn_file, export_settings,
+                     cycle_mat, shader_mat, has_uv):
+    """Export cycles material to godot shader script"""
+    global_script = GlobalScript()
+    vertex_script = VertexShader()
```

Comment:
I would have expected to find the vertex and fragment shaders inside the `GlobalScript()` class. Then the "assemble" function could be part of the GlobalScripts's `to_string()` method.

Reply:
fair enough